pkg/scheduler/factory.go

```diff
@@ -480,40 +473,17 @@ func MakeDefaultErrorFunc(client clientset.Interface, podQueue internalqueue.Sch
 			klog.Errorf("Error scheduling %v/%v: %v; retrying", pod.Namespace, pod.Name, err)
 		}
 
+		if !podQueue.IsPodBeingProcessed(podInfo.Pod.UID) {
```

Comment:
Perhaps this check should be done inside `AddUnschedulableIfNotPresent`, so that we can use the same lock and avoid more racing conditions.

Reply:
SG. I'm going to update along with https://github.com/kubernetes/kubernetes/pull/90660#discussion_r418570480.

Reply:
One glitch here is that existing UTs are quite testing the internal logic of `AddUnschedulableIfNotPresent` without considering the given <pod> is deleted or not, so to keep the UTs still pass, I added a parameter `ensurePodNotDeleted` to this function. PTAL.

Reply:
big no 😅 
Let's update the unit tests instead. There is no point in having logic just for unit tests.

Reply:
> There is no point in having logic just for unit tests.

True. However, existing UTs are all assuming `AddUnschedulableIfNotPresent` can be called arbitrarily - i.e., no need to quite follow the real semantics that it's called during a scheduling cycle upon schedule failure. So I added a helper method `mockSchedCycleWithFailure` in UT to wrap `AddUnschedulableIfNotPresent` for the tests. PTAL at the latest commit.